pkg/metadata/provider_azure.go

```
@@ -0,0 +1,125 @@
+)
+
+const (
+	azMetaDataURL     = "http://169.254.169.254/metadata/instance/"
+	azMetadataVersion = "2017-04-02"
+	azUserDataLoc     = "/var/lib/waagent/ovf-env.xml"
+)
+
+// ProviderAzure is the type implementing the Provider interface for Azure
+type ProviderAzure struct {
+}
+
+// NewAzure returns a new ProviderAzure
+func NewAzure() *ProviderAzure {
+	return &ProviderAzure{}
+}
+
+func (p *ProviderAzure) String() string {
+	return "Azure"
+}
+
+// Probe checks if we are running on Azure
+func (p *ProviderAzure) Probe() bool {
+	// Getting the hostname should always work...
+	_, err := azGet(azMetaDataURL + "compute/name")
+	return (err == nil)
+}
+
+// Extract gets both the Azure specific and generic userdata
+func (p *ProviderAzure) Extract() ([]byte, error) {
+	// Get host name. This must not fail
+	hostname, err := azGet(azMetaDataURL + "compute/name")
+	if err != nil {
+		return nil, err
+	}
+	err = ioutil.WriteFile(path.Join(ConfigPath, Hostname), hostname, 0644)
+	if err != nil {
+		return nil, fmt.Errorf("Azure: Failed to write hostname: %s", err)
+	}
+
+	// public ipv4
+	azMetaGet("network/interface/0/ipv4/ipAddress/0/publicIpAddress", "public_ipv4", 0644)
+
+	// private ipv4
+	azMetaGet("network/interface/0/ipv4/ipAddress/0/privateIpAddress", "local_ipv4", 0644)
+
+	// availability zone
+	azMetaGet("compute/location", "availability_zone", 0644)
+
+	// instance type
+	azMetaGet("compute/vmSize", "instance_type", 0644)
+
+	// instance-id
+	azMetaGet("compute/vmId", "instance_id", 0644)
+
+	// local-hostname
+	azMetaGet("compute/name", "local_hostname", 0644)
+
+	// ssh is not available in Azure
+
+	// Generic userdata, available as XML data
+	userData, err := getUserData()
+	if err != nil {
+		log.Printf("Azure: Failed to get user-data: %s", err)
+		// This is not an error
+		return nil, nil
+	}
+	return userData, nil
+}
+
+// getUserData returns the userdata stored on disk
+func getUserData() ([]byte, error) {
+	return ioutil.ReadFile(azUserDataLoc)
```

sadly there is no URL for fetching userdata.  We can either remove it or let it fail silently if it's not there as I wrote it.